geoips/plugins/modules/readers/generic_satpy.py

```diff
@@ -0,0 +1,134 @@
+particular Satpy reader to be passed via the GeoIPS command line interface
+reader_kwargs argument. The reader_kwargs argument should contain a JSON string
+containing a JSON Object mapping call arguments with their desired values.
+
+In order to map the structure of a Satpy Scene to the Dictionary of Datasets, an
+additional argument named "channel_groups" should be provided via reader_kwargs.
+The argument is a dictionary that groups desired channels by resolution.  This
+dictionary's keys should be strings representing the name of each group and the
+values should be a list of channel names that have the same resolution. The name
+of each group will be used as the keys of the output dictionary and the list of
+channels will be grouped together as a Dataset stored as the associated value to
+the key.  So for example the following channel_groups JSON Object could be used
+to read a selection of visible and IR GOES16 data:
+"channel_groups":{"0.5km":["C02"], "2km":["C13", "C14"]}
+
+While Satpy provides a very large number of readers, a very small number of them
+have been used with this plugin.  This plugin should be considered experimental
+and is not guaranteed to work with all of Satpy's readers.
+"""
+
+import logging
+from typing import Any, Dict, List, Optional, Union
+
+import pyresample.geometry
+import satpy
+import xarray as xr
+
+log = logging.getLogger(__name__)
+
+interface = "readers"
+family = "standard"
+name = "generic_satpy"
+
+
+def call(
+    fnames: List[str],
+    source_name: str,
+    platform_name: str,
+    metadata_only: bool = False,
+    chans: Optional[List[str]] = None,
+    area_def: Optional[pyresample.geometry.AreaDefinition] = None,
+    self_register: bool = False,
+    satpy_reader: str = "abi_L1b",
+    channel_groups: Dict[str, List[str]] = {"0.5km": ["C02"]},
+    roi: float = 3000,
+    filter_parameters: Optional[Dict[str, Any]] = None,
+    satpy_reader_kwargs: Optional[Dict[str, Any]] = None,
+    calibration: Union[List[str], str] = "*",
+    resolution: Union[List[float], float] = "*",
+    polarization: Union[List[str], str] = "*",
+    level: Union[List[float], str] = "*",
+    modifiers: Union[List[str], str] = "*",
+    generate: bool = True,
+    unload: bool = True,
+    load_kwargs: Optional[Dict[str, Any]] = None,
+    force_compute: bool = False,
+) -> Dict[str, xr.DataArray]:
+    if platform_name is None:
+        platform_name = satpy_reader
+
+    scene = satpy.Scene(
+        filenames=fnames,
+        reader=satpy_reader,
+        filter_parameters=filter_parameters,
+        reader_kwargs=satpy_reader_kwargs,
+    )
+
+    all_channels = []
+    for channel_names in channel_groups.values():
+        all_channels.extend(channel_names)
+
+    if load_kwargs is None:
+        load_kwargs = {}
+
+    scene.load(
+        all_channels,
+        calibration=calibration,
+        resolution=resolution,
+        polarization=polarization,
+        level=level,
+        modifiers=modifiers,
```

Comment:
It's possible that these 5 kwargs can be removed. They are used as filters to limit what channels are actually loaded. Or rather, what version of a channel is loaded. For example, if you were loading MODIS L1B data but you wanted the 1km version of all channels even though the 500m of band 1 and 2 was available, you would provide "resolution=[500]" (I think) along with channels being ["1", "2"]. I'm not sure GeoIPS would be able to usefully use this with the dataset naming/grouping later on in this function.

Similarly (alternatively?), it may make sense to first get a list of all DataIDs that the Scene could produce using `scn.available_data_ids()` and then removing any where `data_id["name"]` does not exist in `all_channels`. This would make the use of these 5 filter parameters more useful as this implementation I just described would be loading (potentially) multiple resolutions or calibrations of the same channel. But that would also be needed if someone wanted ABI "C14Rad" versus "C14BT".

Reply:
I think the implementation you described makes sense.  It would be nice to allow people to load multiple calibrations for the same channel.

Reply:
Yes, you're right. There is the Satpy functionality of a `DataQuery` where instead of string names you can pass `DataQuery(name="C01", calibration="radiance")` and other combinations of things that show up in a DataID.

Reply:
Finally getting back to this.  I was originally thinking of altering the code to support allowing multiple calibrations.  However, after evaluating how much time I have to work on this plugin as well as going with the philosophy that this plugin would just address basic usage of Satpy, I'll just go with your initial recommendation of removing the filter kwargs.

Reply:
For the time being, I'm going to leave this one as a possible future improvement/optimization.